openconfig_pyang/plugins/openconfig.py

```diff
@@ -0,0 +1,893 @@
+"""Copyright 2016 The OpenConfig Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+
+Check YANG modules for OpenConfig style and coding conventions.
+
+This checker is derived from the standard pyang lint plugin which also checks
+modules according the YANG usage guidelines in RFC 6087.
+"""
+
+from __future__ import print_function
+
+from enum import Enum
+import optparse
+import os.path
+from pyang import error
+from pyang import plugin
+from pyang import statements
+from pyang.error import err_add
+from pyang.plugins import lint
+import re
+
+from util import yangpath
+
+# Keywords which result in data nodes being created in a YANG tree
+INSTANTIATED_DATA_KEYWORDS = [u"leaf", u"leaf-list", u"container", u"list",
+                              u"choice"]
+LEAFNODE_KEYWORDS = [u"leaf", u"leaf-list"]
+
+
+class ErrorLevel(Enum):
+  """An enumeration of the Pyang error levels.
+
+     - Critical errors are those that are fatal for parsing.
+     - Major errors are used as those that cannot be suppressed, and
+       should result in the module failing submission checks.
+     - Minor errors are used as those that can be suppressed, if there
+       is a clear reason to break a convention.
+     - Warnings are simply statements that the user should be aware of
+       and should not result in submission failures.
+  """
+  CRITICAL = 1
+  MAJOR = 2
+  MINOR = 3
+  WARNING = 4
+
+
+class ModuleType(Enum):
+  """An enumeration describing the type of module.
+
+    OCINFRA: A model that does not need to be validated.
+    NONOC: A non-OpenConfig model.
+    OC: An OpenConfig model.
+  """
+  OCINFRA = 1
+  NONOC = 2
+  OC = 3
+
+
+class ExternalValidationRules(object):
+  required_substatements = {
+      "module": (("contact", "organization", "description", "revision"),
+                 "RFC 6087: 4.7"),
+      "submodule": (("contact", "organization", "description", "revision"),
+                    "RFC 6087: 4.7"),
+      # This rule is explicitly disabled because it is violated by
```

Comment:
might it be better to have a module-level exception for this rather than turn off the check altogether for `iana-if-types`?  The description check in 4.12 is a useful reminder, for example.   Are we checking for these some other way?